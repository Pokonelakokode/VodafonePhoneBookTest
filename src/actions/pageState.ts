import {PageState} from "../reducers/pageState";

export enum PageStateActionTypes {
    SET = 'PAGE_STATE.SET'
}

export type PageStateActions = {
    SET: {
        type: typeof PageStateActionTypes.SET,
        key: keyof PageState,
        data: any
    }
};

export const setPageState = (key: keyof PageState,data:any):PageStateActions['SET'] => ({
    type: PageStateActionTypes.SET,
    key,
    data
});

export const selectUser = (id:number | null):PageStateActions['SET'] => {
    // id ? setHash({user:id}) : setHash();
    return setPageState('selectedUser',id);
};

export type IPageStateActions = PageStateActions[keyof  PageStateActions]