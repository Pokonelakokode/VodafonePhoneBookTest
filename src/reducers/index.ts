import { combineReducers } from "redux";
import pageState from "./pageState";
import users from "./users";

export const rootReducer = combineReducers({
    pageState,
    users
});

export type RootState = ReturnType<typeof rootReducer>