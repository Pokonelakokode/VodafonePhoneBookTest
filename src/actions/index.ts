import {IPageStateActions} from "./pageState";
import {IUserActions} from "./users";
import {RootState} from "../reducers";
import {ThunkAction} from "redux-thunk";
import {Action} from "redux";

export type RootActions = IPageStateActions | IUserActions

export const mainSelector = (state:RootState) => state;

export type EThunkDispatch<A> = ThunkAction<void,RootState,null,Action<A>>