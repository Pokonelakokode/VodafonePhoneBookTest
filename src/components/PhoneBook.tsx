import * as React from "react";
import { useEffect} from "react";
import {getHash} from "../utils";
import {useDispatch, useSelector} from "react-redux";
import {mainSelector} from "../actions";
import {selectUser, setPageState} from "../actions/pageState";
import NewUser from "./NewUser";
import { bindActionCreators } from "redux";
import {router} from "./router";

function PhoneBook() {
    const {pageState:{selectedUser}} = useSelector(mainSelector);
    const actions = bindActionCreators({setPageState,selectUser},useDispatch());
    useEffect(() => {
        const fn = () => actions.setPageState("selectedUser", parseInt(getHash().user) || null);
        window.addEventListener('popstate', fn);
        return () => window.removeEventListener('popstate', fn);
    },[]);
    const [header,content] = router(selectedUser);
    return (
        <div className="container">
            <div className="header row align-items-center">
                {header}
            </div>
            {content}
            <NewUser newUser={() => actions.selectUser(-1)}/>
        </div>
    )
}

export default PhoneBook