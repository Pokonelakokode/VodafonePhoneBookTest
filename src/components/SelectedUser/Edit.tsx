import * as React from "react";
import {useEffect, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import {mainSelector} from "../../actions";
import {addUser, deleteUser, getUserOrNew, saveUser} from "../../actions/users";
import {bindActionCreators} from "redux";
import {selectUser} from "../../actions/pageState";
import {validateEmail} from "../../utils";

interface IProps {
    selectedUser: number
}


const Edit: React.FC<IProps> = ({selectedUser}) => {
    if(selectedUser === null) return <h1>No such user</h1>;
    const {users} = useSelector(mainSelector);
    const actions = bindActionCreators({saveUser, addUser, selectUser, deleteUser}, useDispatch());

    const [user, setUser] = useState(getUserOrNew(users, selectedUser)[0]);
    const [edit, setEdit] = useState(!user.id);
    useEffect(() => {
        let [user, isNew] = getUserOrNew(users, selectedUser);
        setUser(user);
        setEdit(isNew);
    }, [selectedUser]);
    const editCancel = () => {
        if (edit) setUser(getUserOrNew(users, selectedUser)[0]);
        setEdit(!edit);
    };
    const _deleteUser = () => {
        if (confirm(`Are you sure you want to delete ${user.name} from contacts?`)) {
            actions.deleteUser(user.id);
            actions.selectUser(null)
        }
    };
    const nameValid = user.name.length > 0;
    const emailValid = validateEmail(user.email);
    const phoneValid = user.phone.length >= 7;
    const _saveUser = () => {
        let isNew: boolean;
        if(!nameValid || !emailValid || !phoneValid) return alert("Some of your data is invalid");
        [user.id, isNew] = !user.id ? [users.reduce((max: number, _user) => _user.id > max ? _user.id : max, -1) + 1, true] : [user.id, false];
        if (isNew) {
            actions.addUser(user);
            actions.selectUser(user.id);
        } else {
            actions.saveUser(user);
        }
        setEdit(false);
    };

    return (
        <div>
            <h1 className="text-center">{!user.id ? "Add New Contact" : `${edit ? "Edit " : ""}${user.name}'s Profile`}</h1>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Name</label>
                <div className="col-sm-10">
                    <input value={user.name}
                           placeholder="Enter Name"
                           readOnly={!edit} type="text"
                           className={`${nameValid ? "" : "is-invalid "}form-control`}
                           required
                           onChange={(e) => setUser({...user, name: e.target.value})}
                    />
                    {!nameValid && <div className="invalid-feedback d-block">Name is too short</div>}
                </div>

            </div>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Email</label>
                <div className="col-sm-10">
                    <input value={user.email}
                           placeholder="Enter Email"
                           readOnly={!edit} type="email"
                           className={`${emailValid ? "" : "is-invalid "}form-control`}
                           onChange={(e) => setUser({...user, email: e.target.value})}
                    />
                    {!emailValid && <div className="invalid-feedback d-block">Email address is invalid</div>}
                </div>
            </div>
            <div className="form-group row">
                <label className="col-sm-2 col-form-label">Phone</label>
                <div className="col-sm-10">
                    <input value={user.phone} placeholder="Enter Phone"
                           readOnly={!edit} type="text"
                           required
                           className={`${phoneValid ? "" : "is-invalid "}form-control`}
                           onChange={(e) => setUser({...user, phone: e.target.value})}
                    />
                    {!phoneValid && <div className="invalid-feedback d-block">Phone number have to be at least 7 digits</div>}
                </div>
            </div>
            <div className="row justify-content-around">
                {user.id &&
                <button onClick={editCancel} className="btn btn-primary">{edit ? "Cancel" : "Edit"}</button>}
                <button onClick={() => actions.selectUser(null)} className="btn btn-primary">Back</button>
                {edit ?
                    <button onClick={_saveUser} className="btn btn-danger">{user.id ? "Save" : "Create"}</button>
                    :
                    user.id && <button onClick={_deleteUser} className="btn btn-danger">Delete</button>
                }
            </div>

        </div>
    )
};

export default Edit