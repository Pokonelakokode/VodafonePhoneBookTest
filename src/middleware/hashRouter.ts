import {Dispatch, Middleware} from "redux";
import {RootActions} from "../actions";
import {PageStateActionTypes} from "../actions/pageState";
import {setHash} from "../utils";

export const hashRouter:Middleware = () => (next:Dispatch) => (action:RootActions) => {
    next(action);
    if(action.type === PageStateActionTypes.SET && action.key === "selectedUser"){
        action.data ? setHash({user:action.data}) : setHash();
    }
};