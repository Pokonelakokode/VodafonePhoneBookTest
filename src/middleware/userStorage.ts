import {Dispatch, Middleware, MiddlewareAPI} from "redux";
import {RootActions} from "../actions";
import {UserActionTypes} from "../actions/users";
import {saveToStorage} from "../utils";
import {RootState} from "../reducers";

export const userStorage:Middleware = (store: MiddlewareAPI<any,RootState>) => (next:Dispatch) => (action:RootActions) => {
    const {DELETE, ADD, SET} = UserActionTypes;
    next(action);
    if(action.type === DELETE || ADD || SET){
        saveToStorage(store.getState().users)
    }
};

export default userStorage