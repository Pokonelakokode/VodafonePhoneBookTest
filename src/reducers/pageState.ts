import {IPageStateActions, PageStateActionTypes} from "../actions/pageState";
import {getHash} from "../utils";

export type PageState = {
    selectedUser: number | null,
    search: string,
    letterFilter: string
}

const initialState:PageState = {
    selectedUser: parseInt(getHash().user) || null,
    search: "",
    letterFilter: ""
};


export default function pageState(state = initialState, action: IPageStateActions) {
    switch (action.type) {
        case PageStateActionTypes.SET:
            return {...state,[action.key]:action.data};
        default:
            return state
    }
}

