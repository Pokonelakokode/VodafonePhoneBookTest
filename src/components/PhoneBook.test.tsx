import * as React from "react";
import * as renderer from "react-test-renderer"
import {createStore, applyMiddleware} from "redux";
import {rootReducer} from "../reducers";
import {hashRouter} from "../middleware/hashRouter";
import userStorage from "../middleware/userStorage";
import { Provider } from "react-redux";
import PhoneBook from "./PhoneBook";
import {selectUser} from "../actions/pageState";
import { mount } from "enzyme";
import Edit from "./SelectedUser/Edit";


describe("PhoneBook",() => {
    const store = createStore(rootReducer, applyMiddleware(hashRouter, userStorage));
    it('should render snapshot', function () {
        const component = renderer.create(
            <Provider store={store}>
                <PhoneBook/>
            </Provider>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
    it('should render with selectedUser', function () {
        store.dispatch(selectUser(1));
        const wrapper = mount(
            <Provider store={store}>
                <PhoneBook/>
            </Provider>
        );
        expect(wrapper.find(Edit).length).toBe(1);
    });
});