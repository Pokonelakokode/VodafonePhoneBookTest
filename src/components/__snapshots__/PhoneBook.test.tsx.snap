// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneBook should render snapshot 1`] = `
<div
  className="container"
>
  <div
    className="header row align-items-center"
  >
    <div
      className="col-12 col-md-6 pointer-cursor"
    >
      <img
        className="logo"
        onClick={[Function]}
        src="vodafone_logo.svg"
      />
    </div>
    <div
      className="col-12 col-md-6 search-bar"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Search for contact"
        type="text"
        value=""
      />
    </div>
  </div>
  <h2
    className="col-12 text-center"
  >
    Contacts
  </h2>
  <div
    className="col-12 letters"
  >
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      C
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      E
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      G
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      K
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      L
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      M
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      N
    </div>
    <div
      className="letter-selector"
      onClick={[Function]}
    >
      P
    </div>
  </div>
  <div
    className="row"
    id="Contacts"
  >
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Chelsey Dietrich
          </h5>
        </div>
        <div>
          <p>
            Lucio_Hettinger@annie.ca
          </p>
          <p>
            (254)954-1289
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Clementina DuBuque
          </h5>
        </div>
        <div>
          <p>
            Rey.Padberg@karina.biz
          </p>
          <p>
            024-648-3804
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Clementine Bauch
          </h5>
        </div>
        <div>
          <p>
            Nathan@yesenia.net
          </p>
          <p>
            1-463-123-4447
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Ervin Howell
          </h5>
        </div>
        <div>
          <p>
            Shanna@melissa.tv
          </p>
          <p>
            010-692-6593 x09125
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Glenna Reichert
          </h5>
        </div>
        <div>
          <p>
            Chaim_McDermott@dana.io
          </p>
          <p>
            (775)976-6794 x41206
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Kurtis Weissnat
          </h5>
        </div>
        <div>
          <p>
            Telly.Hoeger@billy.biz
          </p>
          <p>
            210.067.6132
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Leanne Graham
          </h5>
        </div>
        <div>
          <p>
            Sincere@april.biz
          </p>
          <p>
            1-770-736-8031 x56442
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Mrs. Dennis Schulist
          </h5>
        </div>
        <div>
          <p>
            Karley_Dach@jasper.info
          </p>
          <p>
            1-477-935-8478 x6430
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Nicholas Runolfsdottir V
          </h5>
        </div>
        <div>
          <p>
            Sherwood@rosamond.me
          </p>
          <p>
            586.493.6943 x140
          </p>
        </div>
      </div>
    </div>
    <div
      className="col-12 col-md-6 col-lg-4 text-center pt-2"
      onClick={[Function]}
    >
      <div
        className="user-card"
      >
        <div
          className="header"
        >
          <h5
            className="col-12"
          >
            Patricia Lebsack
          </h5>
        </div>
        <div>
          <p>
            Julianne.OConner@kory.org
          </p>
          <p>
            493-170-9623 x156
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="pointer-cursor"
    id="NewContact"
    onClick={[Function]}
  >
    +
  </div>
</div>
`;
